### ISSUE TITLE
exitAfterDefer and explicitly cancel()

### ISSUE THREAD

**Issue by Antonboom** [go-critic/go-critic#1022]
Hello!

```golang
# main.go
package main

import (
	"context"
	"os"
)

func main() {
	_, cancel := context.WithCancel(context.Background())
	defer cancel()

	cancel()
	os.Exit(1)
}
```

```
$ gocritic check main.go
./main.go:13:2: exitAfterDefer: os.Exit will exit, and `defer cancel()` will not run
```

**Expected behaviour:**
gocritic does not warn, because I called the `cancel` explicitly before calling `os.Exit`.

**Comment by quasilyte** [comment]
Since this is a simplified example, it's hard to figure out the best way to handle this issue.
gocritic is not wrong that deferred call will not run.
The other call is independent from that deferred call.
In your example, deferred call is redundant and can be removed.

I would personally ask people not to write such code like one in the example.

Instead of this:
```go
func main() {
	_, cancel := context.WithCancel(context.Background())
	defer cancel()
	// main body...
	cancel()
	os.Exit(1)
}
```
It's safer to do this:
```go
func main() {
	mainNoExit()
	os.Exit(1)
}

func mainNoExit() {
	_, cancel := context.WithCancel(context.Background())
	defer cancel()
	// main body...
}
```

**Comment by Antonboom** [comment]
Ok, what do you think about more realistic example?

```golang
func main() {
    ...
    service := app.New(logg, storage)

    server := internalhttp.NewServer(service)
    ctx, cancel := context.WithCancel(context.Background())
    defer cancel()

    go func() {
        signals := make(chan os.Signal, 1)
        signal.Notify(signals, syscall.SIGINT, syscall.SIGHUP)

        <-signals
        signal.Stop(signals)
        cancel()
        // ...
    }()

    logg.Info("service is running...")

    if err := server.Start(ctx); err != nil {
        logg.Error("failed to start http server: " + err.Error())
        cancel()
        os.Exit(1) //nolint:gocritic
    }
}
```

**Comment by quasilyte** [comment]
I see your point, but I would rewrite your main with `mainNoExit` idiom shown above.
Basically, you make all your code exit/fatal-free and return error instead of `log.Fatal`.
Then, in main, you do `log.Fatal` with that error.
This way, there will be 0 conflicts between defer and exit.

```go
func main() {
    // ...
    service := app.New(logg, storage)

    server := internalhttp.NewServer(service)
    ctx, cancel := context.WithCancel(context.Background())
    defer cancel()

    go func() {
        signals := make(chan os.Signal, 1)
        signal.Notify(signals, syscall.SIGINT, syscall.SIGHUP)

        select {
        case <-ctx.Done():
             return
        case <-signals:
        }
        // ...
    }()

    // Start some context bound workers
    // ...

    logg.Info("service is running...")

    if err := server.Start(ctx); err != nil {
        logg.Error("failed to start http server: " + err.Error())
        cancel()
        os.Exit(1) //nolint:gocritic
    }
}
```
=>
```go
func main() {
    if err := mainNoExit(); err != nil {
        logg.Error(err)
    }
}

func mainNoExit() error {
    // ...
    service := app.New(logg, storage)

    server := internalhttp.NewServer(service)
    ctx, cancel := context.WithCancel(context.Background())
    defer cancel()

    go func() {
        signals := make(chan os.Signal, 1)
        signal.Notify(signals, syscall.SIGINT, syscall.SIGHUP)

        select {
        case <-ctx.Done():
             return
        case <-signals:
        }
        // ...
    }()

    // Start some context bound workers
    // ...

    logg.Info("service is running...")

    if err := server.Start(ctx); err != nil {
        return fmt.Errorf("failed to start http server: %w", err)
    }
    return nil
}
```
Note that you don't need to do clumsy direct `cancel()` call.

You may need to pass logg into mainNoExit or use some other logger in `main`, but I think you got the main idea.

Again, I do agree that technically your code may be valid. But gocritic usually report suspicious code, not 100% incorrect code. That code **can** be written in a less error-prone way. I believe that people should focus on that, making code simpler. :)

**Comment by Antonboom** [comment]
@quasilyte, thanks!

I expected you to write an example like this :)

It just looks like a departure from the linter, not like applying some kind of best practice. Although minimizing the number of application exit points is good...

For fun, why is it technically difficult to track something like this?
And I'm ready to close issue.

**Comment by quasilyte** [comment]
> For fun, why is it technically difficult to track something like this?

It's not hard to implement in this restricted case, if we only allow a direct call of deferred expression right before exit/fatal call.

It's harder if we want to solve the general problem of "can we prove that all code paths lead to the deferred expression being called or not"?

We also need to think if we give up on this diagnostic when there is at least 1 direct call to deferred expression or do we try to find all call+exit paths.

Here is an example:

```go
defer cancel()
defer f2()
if cond1 {
  cancel()
  f2()
  os.Exit(1)
}
// ...
if cond2 {
  f2()
  cancel()
  os.Exit(2)
}
```

I would expect that we only allow functions where all deferred expressions are called before exit.
Emphasis on all, as there could be more than one.

> And I'm ready to close issue.

I'm not opposed to adding this specific case as something that gocritic can recognize.

I think it's a matter of 35-50 lines of code (depending on how we count and how many edge-cases we want to cover).
If anyone submits that code, I would happily review it.

So let's not close it. Your point is technically valid and this is a false positive.

**Comment by cristaloleg** [comment]
Is there something left undiscussed?

**Comment by Antonboom** [comment]
@cristaloleg I became a fan of pattern proposed by @quasilyte:
```go
func main() {
    if err := mainImpl(); err != nil {
       log.Fatal(err)
    }
}

func mainImpl() error {
   // ...
}
```

No questions from me more, thanks!

**Comment by cristaloleg** [comment]
Nice! thanks.
